### PR TITLE
LUCENE-9449 Skip docs with _doc sort and "after" (#1725)

### PR DESCRIPTION
**Pull request by mayya-sharipova**
- Enhance DocComparator to provide an iterator over competitive
documents when searching with "after". This iterator can quickly position
on the desired "after" document skipping all documents and segments before
"after".

- Redesign numeric comparators to provide skipping functionality
by default.

Backport for #LUCENE-9449

### CODE REVIEW

**Comment by mayya-sharipova** [comment]
@jimczi  This is a backport for https://github.com/apache/lucene-solr/pull/1725. Not sure if you would like to review it, or I can just merge it.

File: lucene/CHANGES.txt [hunk 1]
```diff
@@ -59,6 +65,12 @@ Optimizations
 * LUCENE-9373: FunctionMatchQuery now accepts a "matchCost" optimization hint.
   (Maxim Glazkov, David Smiley)
 
+* LUCENE-9449: Enhance DocComparator to provide an iterator over competitive
+  documents when searching with "after". This iterator can quickly position
+  on the desired "after" document skipping all documents and segments before
+  "after". Also redesign numeric comparators to provide skipping functionality
+  by default. (Mayya Sharipova, Jim Ferenczi)
```
Comment (jimczi) [inline lucene/CHANGES.txt:72]: The skipping functionality is not by default for numerics in this version ?

**Review by jimczi — APPROVED** [review]
lgtm, I left one comment regarding the description in CHANGES.txt.

File: lucene/CHANGES.txt (hunk 1 above)
Comment (mayya-sharipova) [reply lucene/CHANGES.txt]: Addressed in the last commit

**Pull request merged**